packages/filesystem/src/browser/file-tree/file-tree.ts

```diff
@@ -28,6 +28,50 @@ export class FileTree extends TreeImpl {
     @inject(FileSystem) protected readonly fileSystem: FileSystem;
     @inject(LabelProvider) protected readonly labelProvider: LabelProvider;
 
+    @postConstruct()
+    protected init(): void {
+        this.toDispose.push(
+            this.labelProvider.onDidChange((event: DidChangeLabelEvent) => this.doUpdateElement(event))
+        );
+    }
+
+    protected async doUpdateElement(event: DidChangeLabelEvent): Promise<void> {
+        for (const nodeId of Object.keys(this.nodes)) {
+            const mutableNode = this.nodes[nodeId];
+
+            const nodeWithPossibleUri = mutableNode;
+            if (mutableNode && FileStatNode.is(nodeWithPossibleUri)) {
+                const uri = nodeWithPossibleUri.uri;
+                if (event.affects(uri)) {
+                    mutableNode.name = this.labelProvider.getName(uri);
+                    mutableNode.description = this.labelProvider.getLongName(uri);
+                    mutableNode.icon = await this.labelProvider.getIcon(uri);
+                    this.fireChanged();
```

Comment:
Should be enough to file once after all nodes are processed?

Reply:
> I see that the git history widget respects `onDidChange` event, but the git diff widget does not. Is there a reason on showing old labels for the diff widget?

This is now covered.  Note that we can't simply have the diff label provider 'accepts' function extract the two parts and pass them on.  This would cause infinite recursion.  I have therefore redesigned this so that the two component uris can be obtained and the recursion handled by LabelProvider.

Reply:
We could have a flag `recursion` in `GitLabelProviderContribution` which set to true before calling `LabelProvider` again, so it will skip when `recursion` is true. There is no need to add another API to `LabelProviderContribution`.